File: mobile/lib/modules/workshops/services/workshop_service.dart
/// Servicio de Talleres Flutter (P3 — CU10-CU13).
library;

import '../../../core/api_client.dart';
import '../models/workshop_model.dart';

class WorkshopService {
  // ── CU10 — Listar talleres ────────────────────────────────────────

  static Future<List<Workshop>> listAllWorkshops() async {
    return ApiClient.get<List<Workshop>>(
      '/workshops/all',
      fromJson: (j) => (j as List<dynamic>)
          .map((e) => Workshop.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  static Future<List<Workshop>> listMyWorkshops() async {
    return ApiClient.get<List<Workshop>>(
      '/workshops',
      fromJson: (j) => (j as List<dynamic>)
          .map((e) => Workshop.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  // ── CU FAVORITOS ───────────────────────────────────────────────

  static Future<void> addFavorite(int workshopId) async {
    await ApiClient.post<dynamic>(
      '/workshops/$workshopId/favorite',
      fromJson: (j) => j,
    );
  }

  static Future<void> removeFavorite(int workshopId) async {
    await ApiClient.delete<dynamic>(
      '/workshops/$workshopId/favorite',
      fromJson: (j) => j,
    );
  }

  static Future<List<Workshop>> listMyFavorites() async {
    return ApiClient.get<List<Workshop>>(
      '/workshops/me/favorite-workshops',
      fromJson: (j) => (j as List<dynamic>)
          .map((e) => Workshop.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  static Future<Workshop> registerWorkshop({
    required String nombre,
    required String direccion,
    required double latitud,
    required double longitud,
    String? telefono,
    String? email,
    List<String>? especialidades,
  }) async {
    return ApiClient.post<Workshop>(
      '/workshops',
      body: {
        'nombre': nombre,
        'direccion': direccion,
        'latitud': latitud,
        'longitud': longitud,
        ...?telefono == null ? null : {'telefono': telefono},
        ...?email == null ? null : {'email': email},
        ...?especialidades == null ? null : {'especialidades': especialidades},
      },
      fromJson: (j) => Workshop.fromJson(j as Map<String, dynamic>),
    );
  }

  // ── ESTADO ONLINE Y TÉCNICOS ──────────────────────────────────────

  static Future<bool> getOnlineStatus(int workshopId) async {
    final res = await ApiClient.get<Map<String, dynamic>>(
      '/workshops/$workshopId/online-status',
      fromJson: (j) => j as Map<String, dynamic>,
    );
    return res['en_linea'] as bool? ?? false;
  }

  static Future<List<Technician>> listTechnicians(int workshopId) async {
    return ApiClient.get<List<Technician>>(
      '/workshops/$workshopId/technicians',
      fromJson: (j) => (j as List<dynamic>)
          .map((e) => Technician.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  // ── CU11 — Listar solicitudes ─────────────────────────────────────

  static Future<List<ServiceRequest>> listPendingRequests(int workshopId) async {
    return ApiClient.get<List<ServiceRequest>>(
      '/workshops/$workshopId/requests',
      fromJson: (j) => (j as List<dynamic>)
          .map((e) => ServiceRequest.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  // ── CU12 — Actualizar estado ──────────────────────────────────────

  static Future<ServiceRequest> updateStatus(
    int requestId,
    String estado, {
    String? notas,
    int? tecnicoId,
  }) async {
    return ApiClient.patch<ServiceRequest>(
      '/workshops/requests/$requestId/status',
      body: {
        'estado': estado,
        ...?notas == null ? null : {'notas': notas},
        ...?tecnicoId == null ? null : {'tecnico_id': tecnicoId},
      },
      fromJson: (j) => ServiceRequest.fromJson(j as Map<String, dynamic>),
    );
  }

  // ── CU13 — Historial de atenciones ───────────────────────────────

  static Future<List<ServiceHistory>> getHistory(int workshopId) async {
    return ApiClient.get<List<ServiceHistory>>(
      '/workshops/$workshopId/history',
      fromJson: (j) => (j as List<dynamic>)
          .map((e) => ServiceHistory.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: frontend/src/app/modules/p3_talleres/workshop.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../environment';

// ── Interfaces ──

export interface Workshop {
  id: number;
  usuario_propietario_id: number;
  nombre: string;
  direccion: string;
  latitud: number;
  longitud: number;
  telefono: string | null;
  email: string | null;
  especialidades: string[] | null;
  esta_activo: boolean;
  calificacion_promedio: number;
  creado_en: string;
  en_linea?: boolean;
  estado_registro?: string;
  ultimo_heartbeat?: string;
}

export interface Technician {
  id: number;
  taller_id: number;
  nombre: string;
  telefono: string | null;
  especialidad: string | null;
  esta_disponible: boolean;
  latitud: number | null;
  longitud: number | null;
}

export interface ServiceRequest {
  id: number;
  incidente_id: number;
  taller_id: number;
  tecnico_id: number | null;
  estado: string;
  notas: string | null;
  creado_en: string;
  actualizado_en: string | null;
  // Info Enriquecida
  titulo_incidente?: string;
  categoria_incidente?: string;
  severidad_incidente?: string;
  resumen_ia?: string;
}

export interface ServiceHistory {
  // Base fields from ServiceRequest
  id: number;
  taller_id: number;
  incidente_id: number;
  estado: string;
  notas: string | null;
  creado_en: string;
  actualizado_en: string | null;
  // Info Enriquecida
  titulo_incidente: string | undefined;
  categoria_incidente: string | undefined;
  severidad_incidente: string | undefined;
  resumen_ia?: string;
}

@Injectable({ providedIn: 'root' })
export class WorkshopService {
  private readonly http = inject(HttpClient);
  private readonly base = environment.apiUrl;

  // ── CU10 ──

  registerWorkshop(data: {
    nombre: string;
    direccion: string;
    latitud: number;
    longitud: number;
    telefono?: string;
    email?: string;
    especialidades?: string[];
  }): Observable<Workshop> {
    return this.http.post<Workshop>(`${this.base}/workshops`, data);
  }

  listMyWorkshops(): Observable<Workshop[]> {
    return this.http.get<Workshop[]>(`${this.base}/workshops`);
  }

  listAllWorkshops(): Observable<Workshop[]> {
    return this.http.get<Workshop[]>(`${this.base}/workshops/all`);
  }

  getWorkshop(id: number): Observable<Workshop> {
    return this.http.get<Workshop>(`${this.base}/workshops/${id}`);
  }

  addTechnician(
    workshopId: number,
    data: { nombre: string; telefono?: string; especialidad?: string; latitud?: number; longitud?: number },
  ): Observable<Technician> {
    return this.http.post<Technician>(`${this.base}/workshops/${workshopId}/technicians`, data);
  }

  listTechnicians(workshopId: number): Observable<Technician[]> {
    return this.http.get<Technician[]>(`${this.base}/workshops/${workshopId}/technicians`);
  }

  toggleAvailability(technicianId: number, available: boolean): Observable<Technician> {
    return this.http.patch<Technician>(
      `${this.base}/workshops/technicians/${technicianId}/availability`,
      { esta_disponible: available },
    );
  }

  // ── HEARTBEAT Y PERFIL ──

  getMyProfile(): Observable<Workshop & { tecnicos: Technician[] }> {
    return this.http.get<Workshop & { tecnicos: Technician[] }>(`${this.base}/workshops/me/profile`);
  }

  completeRegistration(): Observable<Workshop> {
    return this.http.post<Workshop>(`${this.base}/workshops/me/complete`, {});
  }

  heartbeat(): Observable<any> {
    return this.http.post(`${this.base}/workshops/heartbeat`, {});
  }

  disconnect(): Observable<any> {
    return this.http.post(`${this.base}/workshops/disconnect`, {});
  }

  // ── CU11 ──

  listPendingRequests(workshopId: number): Observable<ServiceRequest[]> {
    return this.http.get<ServiceRequest[]>(`${this.base}/workshops/${workshopId}/requests`);
  }

  // ── CU12 ──

  updateRequestStatus(
    requestId: number,
    data: { estado: string; notas?: string; tecnico_id?: number },
  ): Observable<ServiceRequest> {
    return this.http.patch<ServiceRequest>(`${this.base}/workshops/requests/${requestId}/status`, data);
  }

  // ── CU13 ──

  getServiceHistory(workshopId: number): Observable<ServiceHistory[]> {
    return this.http.get<ServiceHistory[]>(`${this.base}/workshops/${workshopId}/history`);
  }
}
